### PR TITLE
fix(api): Fix bug where releases with slashes in their name failed to load

### LINKED ISSUE #16012: Decode(?) Sentry release name
## Important Details

How are you running Sentry?

* [ ] On-Premise docker [Version xyz]
* [x] Saas (sentry.io)
* [ ] Other [briefly describe your environment]

## Description

I created a release named 2019-11-04/1, when I try to visit the releases url on sentry.io (sentry.io/organizations/perfectstay/releases/2019-11-04%F4/?project=182xxxx) the following error message is displayed : "This release may not be in your selected project: React"

When I inspect the devtools network panel the GET request returned à 404 not found.

I then decided to replace the / with a "@" and now it is working fine.

## Steps to Reproduce

1. Create a release via init with its name  containing a forward slash (/)
2. try to access this ressource via sentry.io

### What you expected to happen
Access the page

### Possible Solution

Maybe check the decoding part server side ? Or just forbid forward slashes for releases names.

By the way, thank you a lot for your time and for this amazing tool. Love you all. I hope i'll be able to contribute some code this year (:

UPDATE : https://github.com/getsentry/sentry/blob/master/src/sentry/api/endpoints/organization_release_details.py

I think it's in this file, never did some python but I'll try to look at it this week

### PR DESCRIPTION
**Pull request by wedamija**
Not sure this is actually practical in production, because it relies on REQUEST_URI, which I'm not sure will always be available. Also unsure whether using it to overwrite PATH_INFO could have other negative effects.

Possibly a safer option option could be to include the encoded project version in `ReleaseSerializer` as  something like `safeVersion`. Then the FE could just include that as a parameter when making requests to the api, and it'd get double encoded when it encodes the url. Then we can just decode the version like in this diff and it'll work as expected.

Fixes #16012.

### CODE REVIEW

**Comment by wedamija** [comment]
> I'd much rather we lean on some sort of double encoding instead of mutating the wsgi environ dict.

Yeah, I think I agree. Downside is that the URL will be really ugly, but release urls are generally going to be ugly in most cases.

**Comment by mattrobenolt** [comment]
I think the aesthetic of the URL is the least of the concerns imo. :)

**Comment by benvinegar** [comment]
> I think the aesthetic of the URL is the least of the concerns imo. :)

Agreed.

**Comment by markstory** [comment]
Double encoding could work well, how would customers know that they need to double encode? Exposing the doubly encoded version as a property in the serializer would make this and our own front-end easier. 

We will also need to update sentry-cli to do additional encoding.

**Comment by wedamija** [comment]
We'll need follow-up FE diff to start using the new `safeVersion` value when we're passing version as part of the path. If it's in the querystring or a POST payload then the original version should still be used.

Also need to update sentry-cli

File: src/sentry/api/serializers/models/release.py [hunk 1]
```diff
@@ -256,6 +257,7 @@ def serialize(self, obj, attrs, user, *args, **kwargs):
         d = {
             'version': obj.version,
             'shortVersion': obj.short_version,
+            'safeVersion': quote(obj.version, safe=''),
```
Comment (wedamija) [inline src/sentry/api/serializers/models/release.py]: Open to other names for this if anyone has opinions

Comment (markstory) [reply src/sentry/api/serializers/models/release.py]: Some other options could be:

* slug
* urlVersion
* pathName

File: src/sentry/api/utils.py [hunk 2]
```diff
@@ -40,3 +42,18 @@ def get_date_range_from_params(params):
             raise InvalidParams('start must be before end')
 
     return (start, end)
+
+
+def get_release(organization, version, projects=None):
```
Comment (markstory) [inline src/sentry/api/utils.py]: Should this be a class/static method on the Release model? :bike::house:

File: tests/sentry/api/endpoints/test_organization_release_details.py [hunk 3]
```diff
@@ -11,105 +14,153 @@
...
+
+    @fixture
+    def release_1(self):
+        release = Release.objects.create(organization_id=self.organization.id, version='abcabcabc')
+        release.add_project(self.project_1)
+        return release
+
+    @fixture
+    def release_2(self):
+        release = Release.objects.create(organization_id=self.organization.id, version='12345678')
```
Comment (markstory) [inline tests/sentry/api/endpoints/test_organization_release_details.py]: Should one of these releases have a `/` in the version?

File: tests/sentry/api/serializers/test_release.py [hunk 4]
```diff
@@ -453,6 +455,22 @@ def test_release_no_users(self):
         release.add_project(project)
         serialize(release)
 
+    def test_encoded_version(self):
+        """
+        Testing when a repo gets deleted leaving dangling last commit id and author_ids
+        Made the decision that the Serializer must handle the data even in the case that the
+        commit_id or the author_ids point to records that do not exist.
+        """
```
Comment (markstory) [inline tests/sentry/api/serializers/test_release.py]: Is this related to this test?

File: src/sentry/api/serializers/models/release.py (hunk 1 above)
Comment (benvinegar) [reply src/sentry/api/serializers/models/release.py]: Definitely not `safeVersion`

Slug is what we use elsewhere (e.g. `project.slug`).

File: src/sentry/api/utils.py (hunk 2 above)
Comment (wedamija) [reply src/sentry/api/utils.py]: I was thinking about this, but I don't think so. It's really just an API concern, the rest of the codebase doesn't need to know about the double encoding

File: src/sentry/api/serializers/models/release.py (hunk 1 above)
Comment (wedamija) [reply src/sentry/api/serializers/models/release.py]: `slug` sounds like the best option

File: src/sentry/api/utils.py (hunk 2 above)
Comment (wedamija) [reply src/sentry/api/utils.py]: That said, I don't know that utils is a great place either...

Comment (markstory) [reply src/sentry/api/utils.py]: Personally I lean towards 'utils' should never be used, but we're way past that point.

**Comment by benvinegar** [comment]
I'm a little nervous that this is a departure from what we've done before. For example, if this is `slug`, why encode it at all and just strip scary characters and call it a day?

**Comment by wedamija** [comment]
> I'm a little nervous that this is a departure from what we've done before. For example, if this is `slug`, why encode it at all and just strip scary characters and call it a day?

Slug is probably not a perfect name, we could go with something like pathName to be more specific.

One reason to not just strip those characters is so that we can show the release name as the user expects, rather than some munged name. I don't really have enough context to know whether that's important though.

We could just go down the path of adding an actual slug column and doing a data migration as well.

**Comment by mattrobenolt** [comment]
If we strip characters, you can't reverse it back into the database column. So it becomes a lossy process that can't be reversed. Double encoding is reversable at least.

Ideally, you're right, we should store a real slug, but that's gonna be more work than it sounds on the surface. :)

**Comment by wedamija** [comment]
@mattrobenolt @markstory Should we move forward with this then? As is, it's backwards compatible so we can patch the FE and sentry-cli once this is live.

**Comment by markstory** [comment]
I think this moves us towards something better. If this solves the issues customers are facing we could revisit adding a column and filling it with 'clean' data in the future.
